### PR TITLE
[Serialization] Filter internal decls before full deserialization when appropriate

### PR DESCRIPTION
**Pull request by xymus**
Introduce an API to limit which decls to deserialize according to their access level. The filter is applied early, after reading the binary data and before reconstructing the decl. This logic can prevent deserialization failures on reading non-public decls when they're not actually desired.

The current version checks for the simple access level, it doesn't take into account `@usableFromInline` and similar. As such it could be useful to documentation tools, or it would have to be improved to support ABI public decls.

This version applies that filtering only to indexing that already ignores non-public decls after deserializing them.

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@bnbarham What do you think about this version? At this time, it should only apply to users of `ASTWalker::shouldWalkSerializedTopLevelInternalDecls`. The rest is a lot of infrastructure.

**Review by bnbarham — COMMENTED** [review]
I like not deserializing the decls at all when we don't need them 😄. I think my main concern here is that even with this, it's still very easy to introduce problematic calls. It really seems like we need to have the default be safe here with the option to opt into unsafe if you're sure you know what you're doing.

Similar idea for conformances, though that case is a little more hairy as I believe we cache them, so we *have* to deserialize all regardless and then filter later 😞

File: lib/AST/Module.cpp [hunk 1]
```diff
@@ -3033,13 +3035,16 @@ ArrayRef<Decl *> SourceFile::getHoistedDecls() const {
 }
 
 bool FileUnit::walk(ASTWalker &walker) {
+  bool SkipInternal = getKind() == FileUnitKind::SerializedAST &&
+                      !walker.shouldWalkSerializedTopLevelInternalDecls();
```
[the remark this answers is not in the document]
Comment (xymus) [reply lib/AST/Module.cpp:3039]: Never providing internal decls on `getTopLevelDecls` broke only testing tools like sil-opt (that I know of). We can update those, I was keeping it as future work but going for `includeUnsafe` would simply the API so I'll give it a go. Good point.

Other services like `getAllMembers` broke more things and will require more refactoring.

File: include/swift/Sema/IDETypeChecking.h [hunk 2]
```diff
@@ -172,7 +172,7 @@ namespace swift {
   /// \c ModuleDecl::getDisplayDecls() would only return if previous
   /// work happened to have synthesized them.
   void
-  getTopLevelDeclsForDisplay(ModuleDecl *M, SmallVectorImpl<Decl*> &Results, bool Recursive = false);
+  getTopLevelDeclsForDisplay(ModuleDecl *M, SmallVectorImpl<Decl*> &Results, bool recursive = false);
```
[the remark this answers is not in the document]
Comment (xymus) [reply include/swift/Sema/IDETypeChecking.h:175]: Oops, that was unintentional.

**Pull request closed without merging**